// 7441_IC_control/index.js
function seven_segment_display_controller(option){
	this.pinA = option.pin7;
	this.pinB = option.pin1;
	this.pinC = option.pin2;
	this.pinD = option.pin6;
    
	// set all pins to output
	this.setOutput();
	console.log("construct!");
}

seven_segment_display_controller.prototype = {
	
	setOutput : function (){
		this.pinA.rawDirection(true);
		this.pinB.rawDirection(true);
		this.pinC.rawDirection(true);
		this.pinD.rawDirection(true);
	},
    printWord : function (_str){
      if (!isNaN(_str)){
        var _num = parseInt(_str);
        
        if (_num<10){
          var D1 = _num%2;
          _num/=2;
          _num = Math.floor(_num);
          var D2 = _num%2;
          _num/=2;
          _num = Math.floor(_num);
          var D3 = _num%2;
          _num/=2;
          _num = Math.floor(_num);
          var D4 = _num%2;
          
          this.pinA.rawWrite(D1);
          this.pinB.rawWrite(D2);
          this.pinC.rawWrite(D3);
          this.pinD.rawWrite(D4);
        }
        else console.log('error');
      }
      else console.log('over 9');
    }
}
module.exports = seven_segment_display_controller;
